pkg/controller/gitsourceanalysis/gitsourceanalysis_controller.go

```diff
@@ -0,0 +1,152 @@
+	reqLogger.Info("Reconciling GitSourceAnalysis")
+
+	// Fetch the GitSourceAnalysis instance
+	gsAnalysis := &v1alpha1.GitSourceAnalysis{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, gsAnalysis)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		reqLogger.Error(err, "There was an error while reading the GitSourceAnalysis object")
+		return reconcile.Result{}, err
+	}
+
+	if gsAnalysis.Status.Analyzed {
+		reqLogger.WithValues("git-source", gsAnalysis.Spec.GitSourceRef).
+			Info("Skipping GitSourceAnalysis as it was already analyzed")
+		return reconcile.Result{}, err
+	}
+
+	return analyze(reqLogger, r.client, gsAnalysis, request.Namespace)
+}
+
+func analyze(log logr.Logger, client client.Client, gsAnalysis *v1alpha1.GitSourceAnalysis, namespace string) (reconcile.Result, error) {
+	gitSource := &v1alpha1.GitSource{}
+	err := client.Get(context.TODO(), common.NewNsdName(namespace, gsAnalysis.Spec.GitSourceRef.Name), gitSource)
+	if err != nil {
+		gsAnalysis.Status.Error = err.Error()
+		log.WithValues("git-source", gsAnalysis.Spec.GitSourceRef).
+			Error(err, "There was an error while reading the GitSource object")
+
+	} else {
+		buildEnvStats, err := analyzeGitSource(log, client, gitSource, namespace)
+		if err != nil {
+			gsAnalysis.Status.Error = err.Error()
+		} else {
+			gsAnalysis.Status.BuildEnvStatistics = *buildEnvStats
+		}
+	}
+
+	gsAnalysis.Status.Analyzed = true
+	err = client.Update(context.TODO(), gsAnalysis)
+	if err != nil {
+		common.LogWithGSValues(log, gitSource).Error(err, "Error updating GitSourceAnalysis object")
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error updating the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func analyzeGitSource(log logr.Logger, client client.Client, gitSource *v1alpha1.GitSource, namespace string) (*v1alpha1.BuildEnvStats, error) {
+	common.LogWithGSValues(log, gitSource).Info("Analyzing GitSource")
+
+	// Fetch the GitSource secret
+	gitSecret, err := common.GetGitSecret(client, namespace, gitSource)
+	if err != nil {
+		common.LogWithGSValues(log, gitSource, "secret", gitSource.Spec.SecretRef.Name).
+			Error(err, "Error reading the secret object")
+		return nil, err
+
+	} else {
+		var buildEnvStats *v1alpha1.BuildEnvStats
+		var err error
+		if gitSecret != nil {
```

Comment:
it can be nil in the case when the GitSource doesn't contain a secret (it's also covered by test). I agree that it is not obvious from the code - I've changed it a bit, so I hope that it's better now 9c1f609